src/index.js

```diff
@@ -1,74 +1,147 @@
+      password
+    ).then(signed => {
+      return this.executeSignedTransaction(
+        signed.rawTransaction
+      )
+    })
+  }
 
-  @param contractAddress - The address of the contract
-*/
-export function getAllLinks(
-  contractAddress: EthereumAddress,
-): Promise<Array<Link>> {
-  return Promise.resolve([])
-}
+  getAddLinkTransaction(
+    link: Link,
+    draft: TransactionDraft,
+  ): AddLinkTransaction {
+    const data: string = this.instance.methods.addLink(
+      link.url,
+      link.label,
+      link.description,
+      link.getMergedHashtags(),
+    ).encodeABI()
+    return {
+      from: draft.from,
+      to: this.contractAddress,
+      value: 0,
+      gas: draft.gas,
+      gasPrice: draft.gasPrice,
+      data: data,
+      nonce: draft.nonce,
+    }
+  }
 
-/**
-  Get links by hashtag in a specific contract
+  executeSignedTransaction(
+    transaction: SignedTransaction,
+  ): Promise<TransactionReceipt> {
+    return this.web3.eth.sendSignedTransaction(transaction)
+  }
 
-  @param contractAddress - The address of the contract
-  @param hashtag - The hashtag to search links with
-*/
-export function getLinksByHashTag(
-  contractAddress: EthereumAddress,
-  hashtag: HashTag,
-): Promise<Array<Link>> {
-  return Promise.resolve([])
-}
+  getLinksCount(): Promise<number> {
+    return this.instance.methods.getLinksCount().call()
+  }
+
+  getLinkAt(index: number): Promise<?StoredLink> {
+    if (index < 0) {
+      return Promise.reject(
+        new Error('Negative index is not allowed')
+      )
+    }
+    return this.instance.methods.getLinkAt(index).call().then(
+      result => {
+        if (!result.exists) {
+          return null
+        }
+        return new StoredLink(
+          result.url,
+          result.label,
+          result.description,
+          result.hashtags.split('#'),
+          result.timestamp,
+        )
+      }
+    )
+  }
+
+  /**
+    Get all links by a specified address
+
+    @param contractAddress - The address of the contract
+  */
+  getAllLinks(): Promise<Array<StoredLink>> {
+    return this.getLinksSince(0)
+  }
 
-/**
-  Get links submitted by the user
+  /**
+    Get links by a specified address
 
-  @param contractAddress - The addres of the contract
-  @param userAddress - The address of the user
-*/
-export function getLinksByUser(
-  contractAddress: EthereumAddress,
-  userAddress: EthereumAddress,
-): Promise<Array<Link>> {
-  return Promise.resolve([])
+    @param contractAddress - The address of the contract
+  */
+  getLinksSince(
+    since: number,
+  ): Promise<Array<StoredLink>> {
+    return this.getLinksCount()
+      .then(count => {
+        if (since >= count) {
+          return Promise.resolve([])
+        }
+        const indices = [
+          ...Array(count - since).keys(),
+        ].map(ind => ind + since)
+        return Promise.all(indices.map(index => this.getLinkAt(index)))
+      })
+      .then(links => links.filter(Boolean))
```

Comment:
what?

Reply:
Filter non-null values (it's a valid way according to Flow docs)